File: app/src/main/java/com/apps/adrcotfas/goodtime/Statistics/SessionViewModel.java
package com.apps.adrcotfas.goodtime.Statistics;

import android.app.Application;

import com.apps.adrcotfas.goodtime.Database.AppDatabase;
import com.apps.adrcotfas.goodtime.Database.SessionDao;
import com.apps.adrcotfas.goodtime.Session;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

public class SessionViewModel extends AndroidViewModel {

    private SessionDao mSessionDao;
    private ExecutorService mExecutorService;
    private Application mApplication;

    public SessionViewModel(@NonNull Application application) {
        super(application);
        mSessionDao = AppDatabase.getDatabase(application).sessionModel();
        mExecutorService = Executors.newSingleThreadExecutor();
        mApplication = application;
    }

    public LiveData<List<Session>> getAllSessionsByEndTime() {
        return mSessionDao.getAllSessionsByEndTime();
    }

    public LiveData<List<Session>> getAllSessionsByDuration() {
        return mSessionDao.getAllSessionsByDuration();
    }

    public Application getmApplication() {
        return mApplication;
    }

    public LiveData<Session> getSession(long id) {
        return mSessionDao.getSession(id);
    }

    public void addSession(Session session) {
        mExecutorService.execute(() -> mSessionDao.addSession(session));
    }

    public void editSession(Long id, long endTime, long totalTime, String label) {
        mExecutorService.execute(() -> mSessionDao.editSession(
                id, endTime, totalTime, label));
    }

    public void deleteSession(long id) {
        mExecutorService.execute(() -> mSessionDao.deleteSession(id));
    }

    public LiveData<List<Session>> getSessions(String label) {
        return mSessionDao.getSessions(label);
    }

    public LiveData<List<Session>> getAllSessionsUnlabeled() {
        return mSessionDao.getAllSessionsUnlabeled();
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/statistics/SessionViewModel.kt
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.statistics

import com.apps.adrcotfas.goodtime.database.SessionDao
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.apps.adrcotfas.goodtime.database.AppDatabase
import com.apps.adrcotfas.goodtime.database.Session
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.ZoneId
import javax.inject.Inject

@HiltViewModel
class SessionViewModel @Inject constructor(
    database: AppDatabase
) : ViewModel() {

    private val dao: SessionDao = database.sessionModel()

    val allSessions: LiveData<List<Session>>
        get() = dao.allSessions

    val allSessionsUnlabeled: LiveData<List<Session>>
        get() = dao.allSessionsUnlabeled

    val allSessionsUnarchived: LiveData<List<Session>>
        get() = dao.allSessionsUnarchived

    fun getAllSessionsUnarchived(startMillis: Long, endMillis: Long): LiveData<List<Session>>
        = dao.getAllSessionsUnarchived(startMillis, endMillis)

    fun getSession(id: Long): LiveData<Session> {
        return dao.getSession(id)
    }

    fun addSession(session: Session) {
        viewModelScope.launch(Dispatchers.IO) {
            dao.addSession(session)
        }
    }

    fun editSession(session: Session) {
        viewModelScope.launch(Dispatchers.IO) {
            dao.editSession(
                session.id, session.timestamp, session.duration, session.label
            )
        }
    }

    fun editLabel(id: Long?, label: String?) {
        viewModelScope.launch(Dispatchers.IO) {
            dao.editLabel(
                id!!, label
            )
        }
    }

    fun deleteSession(id: Long) {
        viewModelScope.launch(Dispatchers.IO) {
            dao.deleteSession(id)
        }
    }

    fun deleteSessionsFinishedAfter(timestamp: Long) {
        viewModelScope.launch(Dispatchers.IO) {
            dao.deleteSessionsAfter(timestamp)
        }
    }

    fun getSessions(label: String): LiveData<List<Session>> {
        return dao.getSessions(label)
    }
}